dashboard/modules/snapshot/snapshot_head.py

```diff
@@ -212,14 +220,31 @@ async def _get_job_activity_info(self, timeout: int) -> RayActivityResponse:
             )
 
             num_active_drivers = 0
+            latest_job_end_time = 0
             for job_table_entry in reply.job_info_list:
                 is_dead = bool(job_table_entry.is_dead)
                 in_internal_namespace = job_table_entry.config.ray_namespace.startswith(
                     "_ray_internal_"
                 )
+                latest_job_end_time = (
+                    max(latest_job_end_time, job_table_entry.end_time)
+                    if job_table_entry.end_time
+                    else latest_job_end_time
+                )
                 if not is_dead and not in_internal_namespace:
                     num_active_drivers += 1
 
+            current_timestamp = datetime.now().timestamp()
+            # Latest job end time must be before or equal to the current timestamp.
+            # Job end times may be provided in epoch milliseconds. Check if this
+            # is true, and convert to seconds
+            if latest_job_end_time > current_timestamp:
+                latest_job_end_time = latest_job_end_time / 1000
+                assert current_timestamp >= latest_job_end_time, (
```

Comment:
Hm, an assert seems kind of strong. This would cause the endpoint to error, would we want to fail more gracefully instead?

Reply:
The endpoint will return a `RayActivityResponse` with `is_active=RayActivityStatus.ERROR` if this assert statement fails. If the `latest_job_end_time` is not in milliseconds or seconds we could set the `last_activity_at` field in the response to `None`, but it wouldn't be obvious whether this indicates no previous drivers or that this check failed without looking at the dashboard logs. How important do you think it will be to surface this error in the endpoint response?

Reply:
no, setting the result to ERROR sounds good.